File: frontend/main/src/app/service/coach-service.ts
import { ApiResponse } from "../types/api";
import { CoachDataDto, CoachDto, UserCoachDto } from "../types/coach-type";

export const coachService = {
  async getCoaches() {
    const response = await fetch(`/api/coach/slider`, {
      method: 'GET'
    });
    if (!response.ok) throw new Error('Failed to fetch coaches');
    return response.json() as Promise<CoachDto[]>;
  },

  async findCoachById(id: number) {
    const response = await fetch(`/api/coach/data/${id}`, {
      method: 'GET'
    })
    if (!response.ok) throw new Error('Failed to fetch coach');
    return response.json() as Promise<CoachDataDto>;
  },

  async getUserCoach() {
    const response = await fetch(`/api/coach/user`, {
      method: 'GET'
    })
    if (response.status === 401) {
      throw new Error('UNAUTHORIZED');
    }
    if (response.status === 204) {
      throw new Error('NO CONTENT');
    }
    if (!response.ok) throw new Error('Failed to fetch coach');    
    return response.json() as Promise<UserCoachDto>;
  },

  async addUserCoach(coachID: number) {
    const response = await fetch(`/api/coach/add/user/${coachID}`, {
      method: 'POST'
    })
    if (!response.ok) throw new Error('Failed to add coach');
    if (response.status === 401) throw new Error('UNAUTHORIZED');
    return response.json() as Promise<ApiResponse<void>>;
  },
}

File: backend/main-service/src/main/java/main/service/CoachService.java
package main.service;

import main.dto.CoachDataDto;
import main.dto.CoachDto;
import main.dto.UserCoachDto;

import java.util.List;

public interface CoachService {
    //Getting coaches for coach slider
    List<CoachDto> getCoaches();
    //Getting coach data with id
    CoachDataDto getCoachData(long id);
    //Adding coach for user
    void addUserCoach(long userID, long coachID);
    //Getting user coach data can be null
    UserCoachDto getUserCoach(long userID);
}
